packages/contracts/test/attestation.test.ts

```diff
@@ -204,4 +203,80 @@ describe('Attestations', function () {
             .to.emit(unirepContract, 'Attestation')
             .withArgs(epoch, epochKey, attester.address, fieldIndex, val)
     })
+
+    it('should get current attestation counter', async () => {
+        const accounts = await ethers.getSigners()
+        const attester = accounts[1]
+        const epoch = await unirepContract.attesterCurrentEpoch(
+            attester.address
+        )
+
+        const epochKey = BigInt(24910)
+        let attestationCounter = await unirepContract.attestationCounter()
+        expect(attestationCounter).to.equal(0)
+
+        const fieldIndex = SUM_FIELD_COUNT
+        const val = 3
+
+        for (let x = 1; x <= 3; x++) {
+            const tx = await unirepContract
+                .connect(attester)
+                .attest(epochKey, epoch, fieldIndex, val)
+
+            expect(attestationCounter).to.equal(
+                (await unirepContract.attestationCounter()) - 1
+            )
+
+            await expect(tx)
+                .to.emit(unirepContract, 'Attestation')
+                .withArgs(
+                    epoch,
+                    epochKey,
+                    attester.address,
+                    fieldIndex,
+                    BigInt(val) +
+                        (BigInt(attestationCounter) <<
+                            BigInt(254 - REPL_NONCE_BITS))
+                )
+
+            attestationCounter++
+        }
+    })
+
+    it('verify upper bits of replacement field', async () => {
+        const accounts = await ethers.getSigners()
+        const attester = accounts[1]
+        const epoch = await unirepContract.attesterCurrentEpoch(
+            attester.address
+        )
+
+        const epochKey = BigInt(24910)
+        let attestationCounter = await unirepContract.attestationCounter()
+
+        const fieldIndex = SUM_FIELD_COUNT
+
+        for (let x = 1; x <= 3; x++) {
+            const tx = await unirepContract
+                .connect(attester)
+                .attest(epochKey, epoch, fieldIndex, x)
```

Comment:
This is a good test, and I think the logic is right, but it would be nice to attest to a random value. This would guarantee that `x != nonce`, which would cause a false positive.

You can get a random value from `randomf`

```js
import randomf from 'randomf'

const v = randomf(BigInt(2)**BigInt(254-REPL_NONCE_BITS))
```

Reply:
Fixed, using random value for nonce in the test